src/MongoDB.Driver.Core/Core/Authentication/AuthenticationHelper.cs

```diff
@@ -36,28 +35,92 @@ public static void Authenticate(IConnection connection, ConnectionDescription de
             Ensure.IsNotNull(description, nameof(description));
 
             // authentication is currently broken on arbiters
-            if (!description.IsMasterResult.IsArbiter)
+            if (description.IsMasterResult.IsArbiter)
             {
-                foreach (var authenticator in connection.Settings.Authenticators)
-                {
-                    authenticator.Authenticate(connection, description, cancellationToken);
-                }
+                return;
+            }
+
+            var authenticators = connection.Settings.Authenticators.Select((authenticator, index) => new
```

Comment:
Create a sequence tuple of (authenticator, description), where the description has the results of the isMaster SaslAuthMechanisms query.  Named tuples would be nice here, but we haven't upgraded the project to C#7 yet, so I'm using an anonymous Type for now.